src/dotnet/src/Mil.Navy.Nrl.Norm/NormApi.cs

```diff
@@ -0,0 +1,1133 @@
+        /// "repair window" as needed for some applications. While relative paths with respect to the "current working directory"
+        /// may be used, it is recommended that full paths be used when possible.</param>
+        /// <param name="infoPtr">The optional infoPtr and infoLen parameters 
+        /// are used to associate NORM_INFO content with the sent transport object. The maximum allowed infoLen
+        /// corresponds to the segmentSize used in the prior call to NormStartSender(). The use and interpretation of the
+        /// NORM_INFO content is left to the application's discretion.</param>
+        /// <param name="infoLen">The optional infoPtr and infoLen parameters 
+        /// are used to associate NORM_INFO content with the sent transport object. The maximum allowed infoLen
+        /// corresponds to the segmentSize used in the prior call to NormStartSender(). The use and interpretation of the
+        /// NORM_INFO content is left to the application's discretion</param>
+        /// <returns>A NormObjectHandle is returned which the application may use in other NORM API calls as needed.</returns>
+        [DllImport(NORM_LIBRARY)]
+        public static extern long NormFileEnqueue(long sessionHandle, string fileName, byte[]? infoPtr, int infoLen);
+
+        /// <summary>
+        /// This function enqueues a segment of application memory space for transmission within the specified NORM sessionHandle.
+        /// </summary>
+        /// <param name="sessionHandle">Used to identify application in the NormSession.</param>
+        /// <param name="dataPtr">The dataPtr parameter must be a valid pointer to the area of application memory to be transmitted.</param>
+        /// <param name="dataLen">The dataLen parameter indicates the quantity of data to transmit.</param>
+        /// <param name="infoPtr">The optional infoPtr and infoLen parameters 
+        /// are used to associate NORM_INFO content with the sent transport object. The maximum allowed infoLen
+        /// corresponds to the segmentSize used in the prior call to NormStartSender(). The use and interpretation of the
+        /// NORM_INFO content is left to the application's discretion.</param>
+        /// <param name="infoLen">The optional infoPtr and infoLen parameters 
+        /// are used to associate NORM_INFO content with the sent transport object. The maximum allowed infoLen
+        /// corresponds to the segmentSize used in the prior call to NormStartSender(). The use and interpretation of the
+        /// NORM_INFO content is left to the application's discretion</param>
+        /// <returns>A NormObjectHandle is returned which the application may use in other NORM API calls as needed.</returns>
+        [DllImport(NORM_LIBRARY)]
+        public static extern long NormDataEnqueue(long sessionHandle, byte[] dataPtr, int dataLen, byte[]? infoPtr, int infoLen);
```

Comment:
A managed buffer won't work so well here. Managed data can move in memory and NORM needs to reference the data after the call to Enqueue. Forcing the users of this wrapper to do the pinning seems harsh :) One option is to have users use unmanaged memory by passing an IntPtr or byte*. Another would be to have the wrapper handle the pinning/unpinning transparently.

Reply:
@lbargaoanu Please review updated `NormDataEnqueue` `extern` method and wrapper method to verify that the new approach addresses the comment.

Reply:
I'm afraid that won't really work. It would work if NORM were to copy the user buffer inside Enqueue. But it doesn't, it stores the pointer and reads from it when it needs to. So it's safe to unpin the buffer only when we know that NORM doesn't need it anymore. When the object is purged, when the send is cancelled and when the sender is stopped. Not sure whether the purged event is fired when the sender is stopped or if there are other interesting cases.

Reply:
Maybe cancel triggers the object purged notification:

https://github.com/USNavalResearchLaboratory/norm/blob/2ae69e6e96d1ea5dab9d8353d2b8ca80b3a7683b/src/common/normApi.cpp#L2697

Reply:
Actually it seems to trigger NORM_RX_OBJECT_ABORTED.

Reply:
@lbargaoanu  Thanks for the clarification. Since the buffer cannot be unpinned directly after the the `NormDataEnqueue` call, a static wrapper in method in `NormApi` is probably not the best fit. The wrapper would probably work better as part of `NormSession` and `NormObject` since they are storing session and object state. The `NormSession` `DataEnqueue` method can create the pinned handle. The handle can be stored with `NormObject` and then freed when one the events occur that indicates that NORM does not need the buffer.

Reply:
Yes, smth like that.

Reply:
After trying different options for the `NormDataEnqueue` buffer, letting the user handle the lifetime of the buffer seems to be the better option than trying to wrap the pinning/unpinning transparently. 

To summarize the options that were tried for `NormDataEnqueue` buffer:

- Wrap the pinning/unpinning transparently: Initially seemed like the best option, but did not fit the way events are handled. The user checks for events and would require significant redesign for the .NET extension to track the buffer and automatically unpin based on events.
- Model Java extension use of `ByteBuffer`: The Java extension `NormSession` uses `ByteBuffer` for `dataEnqueue()` instead of `byte` array. The Java `NormMsgr` example uses `ByteBuffer.allocateDirect()` to create a `DirectByteBuffer` which is allocated in native memory outside of the Java heap. The Java `NormMsgr` example then uses `ByteBuffer.put()` to copy an existing `byte` array to the direct buffer. .NET does not have a direct equivalent of `DirectByteBuffer`, but a similar class can be defined that uses `Marshal` to implement equivalent functionality. This would provide a consistent API for Java and .NET. However, this does seem to force a Java way of doing something in .NET.
- Wrap pinning/unpinning logic: Wrap the pinning/unpinning logic in a class so that the user can pin/unpin a managed buffer without having to write all of the pinning/unpinning logic. The wrapper class can implement `IDisposable`, which is the standard .NET interface for releasing unmanaged resources. The user can implemement a `using` statement or block if the NORM event handling is in the same method that the pinned buffer is created. If NORM event handling is in a separate method, the `Dispose()` method can be called directly. Also, an extension method can be used to further simplify the process so user can call a method on the managed buffer: `managedBuffer.ToPinnedObject()`. This seems to be the most direct solution compared to the other options and also seems to be more of a .NET approach compared to the `ByteBuffer` option.

Finishing a feature branch for wrapping the pinning/unpinning logic and then will update the main branch.

Would also like to try out any other options not listed for the `NormDataEnqueue` buffer.

Reply:
I see your points, but solution no 3, besides the usability issues, also has performance problems. As I've mentioned before, the buffers will be most likely pinned for a long time and that will lead to heap fragmentation.and slower gc. One way arround it would be to use the pinned object heap, with `Gc.AllocateArray`. This would work best with pooling, but there is no built-in implementation. Without pooling, it would require some wrapper class to enforce proper usage. And the user would still need to create and keep track of the buffers.
So perhaps an unmanaged buffer (maybe a `SafeBuffer`) would work best. It would make the user's job a little more difficult, but it seems less problematic over all. You wouldn't have to enforce some way to work with the unmanaged memory, but you can show one way to do it, perhaps with `UnmanagedMemoryAccessor` or `UnmanagedMemoryStream`.

Reply:
@lbargaoanu Thanks for the feedback. `SafeBuffer` seems to be the best solution for `NormSession.DataEnqueue()`. It solves the pinning/unpinning issue and uses a .NET Standard class that is fairly consistent with the Java extension. `NormApi.NormDataEnqueue()` can use `IntPtr` and `NormSession.DataEnqueue()` can use built-in `SafeBuffer` functionality to obtain the pointer to pass to `NormApi.NormDataEnqueue()`.

Can include an example of using .NET Standard classes for working with unmanaged memory,`MemoryMappedFile` and `SafeMemoryMappedViewHandle`, and passing the `SafeBuffer` to `NormSession.DataEnqueue()`. Can also add an example of using a custom class derived from `SafeBuffer` that can be used similar to Java `ByteBuffer`.

Reply:
I didn't notice that all the generic `SafeBuffer` implementations are internal. `SafeMemoryMappedViewHandle` is too specialized and we don't need to bring memory mapped files in the conversation :) So maybe just stick with a raw pointer. In the end, the user doesn't have to keep track of this buffer, NORM can [provide](https://github.com/USNavalResearchLaboratory/norm/blob/669fc4f6719a3f05869f7adec4b3a1c5329d007b/examples/normDataSend.cpp#L207) it.

Reply:
Maybe having both is best. For example, `UnmanagedMemoryStream` has constructors for both. The thing is, when using a `SafeBuffer`, the user has to keep a reference to a buffer, otherwise it could be finalized and the memory freed while NORM is still using it. You can work with a pointer without keeping track because you can get it from NORM.

Reply:
Good point about having both. Can just overload the `NormSession.DataEnqueue()` method.

Reply:
@lbargaoanu Pushed updates based on the conversation. Please verify if the comments have been addressed and if there are other recommendations.

Reply:
The initial problem is solved.